controller/instance_manager_controller.go

```diff
@@ -515,6 +520,130 @@ func (imc *InstanceManagerController) annotateCASafeToEvict(im *longhorn.Instanc
+		}
+		if err != nil {
+			return false, false, false, err
+		}
+		if !isSettingSynced {
+			return false, false, false, nil
+		}
+	}
+
+	return true, false, false, nil
+}
+
+func (imc *InstanceManagerController) isSettingTaintTolerationSynced(setting *longhorn.Setting, pod *corev1.Pod) (bool, error) {
+	newTolerationsList, err := types.UnmarshalTolerations(setting.Value)
+	if err != nil {
+		return false, err
+	}
+	newTolerationsMap := util.TolerationListToMap(newTolerationsList)
+	lastAppliedTolerations, err := getLastAppliedTolerationsList(pod)
+	if err != nil {
+		return false, err
+	}
+
+	return reflect.DeepEqual(util.TolerationListToMap(lastAppliedTolerations), newTolerationsMap), nil
+}
+
+func (imc *InstanceManagerController) isSettingNodeSelectorSynced(setting *longhorn.Setting, pod *corev1.Pod) (bool, error) {
+	newNodeSelector, err := types.UnmarshalNodeSelector(setting.Value)
+	if err != nil {
+		return false, err
+	}
+	if pod.Spec.NodeSelector == nil && len(newNodeSelector) == 0 {
+		return true, nil
+	}
+
+	return reflect.DeepEqual(pod.Spec.NodeSelector, newNodeSelector), nil
+}
+
+func (imc *InstanceManagerController) isSettingGuaranteedInstanceManagerCPUSynced(setting *longhorn.Setting, pod *corev1.Pod) (bool, error) {
+	lhNode, err := imc.ds.GetNode(pod.Spec.NodeName)
+	if err != nil {
+		return false, err
+	}
+	if types.GetCondition(lhNode.Status.Conditions, longhorn.NodeConditionTypeReady).Status != longhorn.ConditionStatusTrue {
+		return true, nil
+	}
+
+	resourceReq, err := GetInstanceManagerCPURequirement(imc.ds, pod.Name)
+	if err != nil {
+		return false, err
+	}
+	podResourceReq := pod.Spec.Containers[0].Resources
+	return IsSameGuaranteedCPURequirement(resourceReq, &podResourceReq), nil
+}
+
+func (imc *InstanceManagerController) isSettingPriorityClassSynced(setting *longhorn.Setting, pod *corev1.Pod) (bool, error) {
+	return pod.Spec.PriorityClassName == setting.Value, nil
+}
+
+func (imc *InstanceManagerController) isSettingStorageNetworkSynced(setting *longhorn.Setting, pod *corev1.Pod) (bool, error) {
+	nadAnnot := string(types.CNIAnnotationNetworks)
+
+	return pod.Annotations[nadAnnot] == setting.Value, nil
+}
+
+func (imc *InstanceManagerController) isSettingDataEngineSynced(settingName types.SettingName, im *longhorn.InstanceManager) (bool, error) {
+	enabled, err := imc.ds.GetSettingAsBool(settingName)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get %v setting for updating data engine", settingName)
+	}
+	var dataEngine longhorn.DataEngineType
+	switch settingName {
+	case types.SettingNameV1DataEngine:
+		dataEngine = longhorn.DataEngineTypeV1
+	case types.SettingNameV2DataEngine:
+		dataEngine = longhorn.DataEngineTypeV2
+	}
+	if !enabled && im.Spec.DataEngine == dataEngine {
+		imc.logger.Infof("Cleaning up the instance manager %v for %v data engine", im.Name, dataEngine)
+		if err := imc.ds.DeleteInstanceManager(im.Name); err != nil {
+			return false, err
+		}
+	}
```

Comment:
We should not delete instance managers in `isSettingDataEngineSynced`. This method is used for checking if the setting synced or not instead fo doing any action.

Reply:
Removed
```golang
imc.logger.Infof("Cleaning up the instance manager %v for %v data engine", im.Name, dataEngine)
if err := imc.ds.DeleteInstanceManager(im.Name); err != nil {
    return false, err
}
```
and used instead
```golang
return false, nil
```